### DISCUSSION TITLE
Exception Thrown Using Step Labels

### DISCUSSION THREAD

**Discussion by Dmo16** [Gremlinq/ExRam.Gremlinq#345] [Q&A] [answered]
Placing this here instead of a bug report at the moment b/c I'm erring on the side that I'm doing something wrong, vs an issue with the code.

Following this example code:
![image](https://user-images.githubusercontent.com/45639728/115971814-aa4e7a00-a518-11eb-8494-eb1dade9e63e.png)

I wrote a method in similar style:
```

                var team = await g
                    .V<RaceTeam>()
                    .As((__, raceTeam) => __
                        .Out<OwnedByUser>()
                        .OfType<AimUser>()
                        .As((__, user) => __
                            .Select(raceTeam, user)));
```

However, the first `.As` statement throws a StepLabel exception: 

```
System.NotImplementedException: The conversion operator on StepLabel is not intended to be called. It's use is to appear in expressions only.
   at ExRam.Gremlinq.Core.StepLabel`1.ThrowConversion() in /_/src/ExRam.Gremlinq.Core/Queries/StepLabel.cs:line 72
   at ExRam.Gremlinq.Core.StepLabel`1.get_Value() in /_/src/ExRam.Gremlinq.Core/Queries/StepLabel.cs:line 67
```

I'm _fairly_ certain this code should work, but I'm all ears for what I messed up! :D

**Comment by danielcweber** [comment]
Hard to tell without a minimal executable repro. Something's calling 'Value' on the step label, it doesn't appear in the posted code snippet though.

**Reply by Dmo16** [reply]
Hmm, the only other thing that is dealing with the results is this: 

`var result = _mapper.Map<RaceTeam>(team.Item1);`

and then this: 

```
result.Owner = team.Item2;

return result;
```

Perhaps I am misunderstanding the encapsulation here but, shouldn't the only things trying to call anything inside the steps be related to GremLinQ itself, or am I misunderstanding what the `.Select(raceTeam, user)` statement is supposed to do?

Based on the description from your docs I _thought_ it would return tuples I could work with, but maybe I'm not using them correctly?

**Reply by danielcweber** [reply]
Again, very hard to tell what's going on here without a repro that I can just compile and run.

**Reply by Dmo16** [reply]
I understand. 

I didn't think a repo was necessary to answer whether or not the `.Select` statement is meant to return tuples that can be assigned, parsed, read, etc.

The code attempting to use these tuples is posted above.

I'll keep poking around both of our code bases and see if I can isolate it.

Thank you for your time!

**Comment by danielcweber** [comment]
>Based on the description from your docs I thought it would return tuples I could work with,

Sure they are supposed to

>but maybe I'm not using them correctly?

and that's what I'm trying to figure out, because it's not obvious, but I can't without a minimal repro.

**Comment by danielcweber** [comment, marked as answer]
Answered in ##376
